### ISSUE TITLE
Support webm format.

### ISSUE THREAD

**Issue by tan-tan-kanarek** [kaltura/nginx-vod-module#1231]
Hi,
Transcoding webm files (vp9 + opus) to mp4 files causing them to lose their transparency in chrome.
I wonder what's the effort to support webm and if you could direct me what should be done.
The features I'll need include transparency, sequence of several files and clipping, with few alternative audio tracks.

**Comment by erankor** [comment]
VP9 and Opus can live inside MP4 container, last time I checked, it was experimental in ffmpeg and required some flag to enable, not sure what's the current status... So you should be able to just repackage with `-c:v copy -c:a copy`.
Note that VP9 is supported only in DASH, and probably plays only in Chrome.

**Comment by tan-tan-kanarek** [comment]
Indeed VP9 in MP4 works, with opus (-strict -2), but the transparency lost, no matter if I convert it from WEBM or build it from transparent PNG frames.
With MP4 I can't change the background behind the video while with WEBM I can, as video element source or as dash manifest using dash.js player.
I understand the limitation of dash only, but I wasn't sure if it should be simple enough to clip WEBM.
Alternatively, because chrome is efficient even with large WEBM files, I can build the dash manifest manually myself, but I still would like to clip the different source dynamically, maybe using mapped or parameterized progressive download.

**Comment by erankor** [comment]
When you wrote that it doesn't work with MP4 you meant using the module to generate DASH or the MP4 directly? 
When the module generates a DASH stream with VP9 codec, it outputs webm container not fMP4 (regardless of how the input was supplied).
If this indeed doesn't work, maybe it's worth comparing the webm fragments created by the module to the original webm in which transparency works, maybe some webm tag gets lost there.
You can also try to compile the module with webm support, by uncommenting this line -
https://github.com/kaltura/nginx-vod-module/blob/master/ngx_http_vod_module.c#L256
But in general, webm (mkv) support is in POC level, will require a significant effort IMHO to get it to work well.

**Comment by tan-tan-kanarek** [comment]
Already tried building it with mkv support, it was early in the day, can't remember what failed, however, I tried MP4 with vp9 directly or with HLS, I didn't test them with DASH.
I tested DASH with WEBM and MKV inputs, which failed.
Will test and update.
Thank you.

**Comment by tan-tan-kanarek** [comment]
OK, so as I said, it doesn't work with MP4.

Source file: [chapter-denoised.mp4](https://drive.google.com/file/d/1oohpmlinx9uMGSn9_zGtnmsM3f-Q7Ebp/view?usp=sharing)
I created PNG frames with transparency: [mask-resized.rar](https://drive.google.com/file/d/1WEmsCOyITK4kQJ1Ur7SlnUNVgz0eznXF/view?usp=sharing)

From them I created WEBM: [mask-resized.webm](https://drive.google.com/file/d/1zTV37Gj3ZICoycbfwazNE2Cq1wmR-6a-/view?usp=sharing)
`ffmpeg -i .\chapter-denoised.mp4 -r 30 -i .\mask-resized\%06d.png -c:v libvpx-vp9 -vf fps=30 -c:a opus -strict -2 -map 0:a:0 -map 1:v -y mask-resized.webm`

Converted to MP4: [mask-resized.webm.mp4](https://drive.google.com/file/d/1zE4oEkbvf7AestYh-PwXVQhr2SFf4oM3/view?usp=sharing)
`ffmpeg -i mask-resized.webm -c:a copy -c:v copy -strict -2 -y mask-resized.webm.mp4`

Also tried to create MP4 directly from PNG: [mask-resized.mp4](https://drive.google.com/file/d/1c68ajCedaLI6dQt8j2lZ5p77VDc5EKM-/view?usp=sharing)
`ffmpeg -i .\chapter-denoised.mp4 -r 30 -i .\mask-resized\%06d.png -c:v libvpx-vp9 -vf fps=30 -c:a opus -strict -2 -map 0:a:0 -map 1:v -pix_fmt yuva420p -crf 30 -y mask-resized.mp4`

[DASH resulted manifest and segments](https://drive.google.com/file/d/1bqoGInFYkx7I542tODGMu-lgTWi55kvn/view?usp=sharing).

The JSON manifest is simple:
```
{
    "durations": [12930],
    "sequences": [{
            "label": "original",
            "clips": [{
                    "path": "/segments/test/11/mask-resized.webm.mp4",
                    "tracks": "v1-a1",
                    "type": "source",
                    "clipTo": 12.93,
                    "clipFrom": null
                }
            ]
        }
    ]
}
```

**Comment by tan-tan-kanarek** [comment]
Important info:
I encountered the same issue when I created manifest using ffmpeg:
`ffmpeg -i mask-resized.webm -c copy -dash_segment_type webm -f dash mask-resized.mpd`

However, 
When I used vp9 both in the encoder (copy in my case) and the **decoder**, it solved the problem:
`ffmpeg -c:v libvpx-vp9 -i mask-resized.webm -c copy -dash_segment_type webm -f dash mask-resized.mpd`

**Comment by tan-tan-kanarek** [comment]
Another interewsting finding that can promote this.
I created mp4 from png files using png as codec:
`ffmpeg -i .\chapter-denoised.mp4 -r 30 -i .\mask-resized\%06d.png -c:v png -vf fps=30 -c:a aac -strict -2 -map 0:a:0 -map 1:v -pix_fmt yuva420p -crf 30 -y mask-resized.png.mp4`

Later I created mpd from this mp4 without specifying decoder:
`ffmpeg -i mask-resized.png.mp4 -c:v libvpx-vp9 -c:a opus -dash_segment_type webm -f dash mask-resized.mpd`

It worked with transparency.
Which means, I can use mp4 files, and I don't need to add support for webm, but I will need to add support for png codec which I presume will be easier.

**Comment by erankor** [comment]
You are transcoding to vp9 in the last command, the dash stream doesn't have png, so don't see how png will help...
But - do you have 2 dash streams, one that works with transparency and one that doesn't? If so, can you send URLs?

**Comment by tan-tan-kanarek** [comment]
PNG codec helps as interim codec for the MP4 because VP9 on MP4 loses its transparency while PNG preserve it.
Transcoding MP4-VP9 to WEBM-VP9 won't restore transparency while MP4-PNG will.
Will send the streams privately.

**Comment by erankor** [comment]
I looked at the samples, the working sample has in the init segment `AlphaMode 1` while the non-working one has a few other tags - `FlagInterlaced 2, Colour => Range 1`. 
From a quick search in ffmpeg code, I think the alpha data goes nowhere - there are references to AV_PKT_DATA_MATROSKA_BLOCKADDITIONAL in vpx encoder/decoder and in matroska encoder/decoder but nowhere else.
The "spec" for VP9/MP4 has no mention of alpha - https://www.webmproject.org/vp9/mp4/.
So, I guess it is indeed not possible to encode the alpha channel in MP4 container.

**Comment by tan-tan-kanarek** [comment]
Thanks,
I guess that adding support for WEBM will take more effort, but adding support for PNG and on-the-fly transcoding to VP9 will be less efficient at run time.

**Comment by erankor** [comment]
Yes, also if you do VP9 transcoding on-the-fly, you will have to explicitly handle the alpha data, as the ffmpeg matroska encoder does.
